### PR TITLE
Fix code lens count to match reference list (issue #637)

### PR DESCRIPTION
**Pull request by bitwisecook**
## Summary

Fix issue #637 where the code lens reference count above a proc definition could disagree with the actual reference list shown when clicking the lens or running Find All References.

The root cause was that the count came from `proc_usage_counts` (a workspace-level aggregate keyed by resolved qualified names), while the reference list came from name-based resolution. Forward references and cross-file calls have no resolved name at analysis time, so they were missing from the count but still matched by the reference resolver, causing "0 references" to appear above clearly-used procs.

The fix makes `resolve_code_lens` derive the count from the same reference resolution that backs the peek, ensuring they can never drift. The cached `proc_usage_counts` is retained as a fallback for callers (like unit tests) that don't supply a resolver.

## Changes

- **server/features/code_lens.py**: Modified `resolve_code_lens` to compute the count from `find_references` when available, falling back to `proc_usage_counts` only when no resolver is supplied
- **tests/test_code_lens.py**: Added comprehensive regression tests covering forward references, qualified calls, namespace calls, and other edge cases that previously diverged
- **docs/kcs/kcs-issue-code-lens-zero-references.md**: Added user-facing KCS note explaining the symptom, fix, and root cause
- **docs/kcs/README.md**: Linked the new KCS note

## Validation

- Added `test_count_derives_from_resolved_locations` to verify the count matches the resolver output
- Added `TestCountMatchesReferences` class with `test_count_equals_reference_list` covering 7 different call patterns (forward refs, qualified calls, namespace calls, etc.)
- Added `test_forward_reference_is_not_zero` to specifically verify the headline symptom is fixed
- All new tests pass; existing tests remain unaffected

https://claude.ai/code/session_01RTrrHV2RAfjADrYyNRwZWV

### CODE REVIEW

**Comment by bitwisecook** [comment]
**The Codex P2 finding checks out — and it points at `find_proc_call_sites`, not the lens.**

I reproduced the scenario against `find_proc_call_sites` on this branch. For an unresolved call (a forward reference, or a cross-file call) the resolver scopes nothing:

```python
if resolved_name is not None:
    matches_target = resolved_name == qualified_name
else:
    matches_target = invocation.name in call_forms   # {name, qualified_name, qualified_no_prefix}
```

With `foo` called before its definition inside `namespace eval a`, and both `::a::foo` and `::b::foo` defined, the call has no `resolved_qualified_name`, so `invocation.name == "foo"` lands in `call_forms` for **both** procs — and `::b::foo` reads `1 reference` for a call that isn't its.

One nuance worth flagging: because this same `find_proc_call_sites` backs both the resolve-count and the peek, the count and the peek still **agree**, so #644's headline "they can't drift" goal holds. The residual defect is purely a *correctness* one in the shared resolver, which mis-attributes the unresolved call to every same-named proc. So the fix belongs in `find_proc_call_sites` (scope an unresolved call to the namespace it sits in, or reject ambiguous tail-name matches) — once that lands, both the count and the peek become correct for free.

FWIW the Rust port derives the lens count from its namespace-aware reference resolver (`call_ns == target_ns` for the simple/unresolved-name case), so it already attributes the forward call only to `::a::foo` and leaves `::b::foo` at `0 references`. That scoping rule could be a useful reference for the Python side.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/features/code_lens.py [hunk 1]
```diff
@@ -101,10 +104,23 @@ def resolve_code_lens(
...
+        # parallel ``proc_usage_counts`` aggregate. The two used to drift
+        # (issue #637): a forward or cross-file call resolves to ``None`` at
+        # analysis time, so it is absent from ``proc_usage_counts`` (keyed by
+        # the resolved qualified name) yet still matched by name in the
+        # reference resolution — giving "0 references" above a proc the peek
+        # listed as used. ``proc_usage_counts`` remains the fallback only for
+        # callers (e.g. unit tests) that supply no resolver.
+        if find_references is not None:
+            locations = find_references(data.uri, data.qname)
+            count = len(locations)
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply server/features/code_lens.py:118]: Good catch — this was a real pre-existing over-attribution in `find_proc_call_sites` (the peek/Find All References already credited the bare call to both procs); deriving the count from that resolution surfaced it in the title too.

Fixed by rejecting the ambiguous bare-name match: when several procs share a simple name (e.g. `::a::foo` and `::b::foo`), an *unresolved* bare `foo` call can't be attributed to a specific one without the call site's namespace — which `CommandInvocation` doesn't carry — so the bare-name fallback is dropped for that case. Resolved calls and qualified call forms (`a::foo`) stay precise, and the count and the peek remain consistent (both now show 0 on the ambiguous call rather than a phantom 1 on the unrelated proc).

Verified on your exact example:

```tcl
namespace eval a { foo; proc foo {} {} }
namespace eval b { proc foo {} {} }
```

`::a::foo` and `::b::foo` both now report `0 references` instead of `::b::foo` showing `1`. Added a regression test (`test_ambiguous_simple_name_not_double_counted`) covering it. The proper fix for attributing the call to the correct namespace would need scope context threaded onto `CommandInvocation`, which is a larger change tracked separately.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bitwisecook** [comment]
**Note for the record — Rust port has caught up to this fix.**

The native Rust LSP now carries the equivalent of #644 (on `claude/intelligent-thompson-48xxbh`):

- The code-lens proc count is derived from the same namespace-aware `references` resolver that backs Find All References / the peek, so the title and the peek can't drift — the structural guarantee of this PR.
- Regression tests cover the #644 call patterns (forward ref, qualified / namespace-qualified calls, body call, command substitution, unused proc), asserting `count == len(references)`.
- It also pins the ambiguous-forward-ref case from the review above: `foo` called before its definition in `namespace eval a`, with `::a::foo` and `::b::foo` both defined, is credited only to `::a::foo` (`::b::foo` stays `0 references`). This is the case the Python `find_proc_call_sites` fallback still mis-attributes, so the two implementations will diverge on it until the Python resolver scopes unresolved calls by namespace.

No action needed on this PR for the Rust side; flagging only so the cross-implementation status is tracked here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bitwisecook** [comment]
Adopted the namespace-aware approach you pointed at from the Rust port, rather than the blunt "reject ambiguous" option. `find_proc_call_sites` now resolves an unresolved bare call the way Tcl does (current namespace, then global), so the forward `foo` in `namespace eval a` is credited to `::a::foo` and `::b::foo` stays at 0 — matching `call_ns == target_ns`. `CommandInvocation` carries the call site's enclosing namespace (set from the analyser scope chain, propagated through `conf_wrapped`).

One nuance your comment helped surface: disambiguation only kicks in when the simple name is ambiguous *within the analysed document*. An earlier cut keyed off "does the proc exist locally", which dropped genuine **cross-file** references (the proc lives in another document, so it isn't in this file's `all_procs`) and regressed cross-document call hierarchy. So the rule is: ambiguous-within-this-file → namespace-resolve; unique or absent → name-match (preserves cross-file). Full suite is green (17946 passed).

Test coverage at all three layers, as requested:
- unit (`tests/test_code_lens.py`): `::a::foo` = 1, `::b::foo` = 0;
- LSP e2e (`tests/lsp_e2e/test_editor_features_e2e.py`): resolves the lens over JSON-RPC against the packaged server and asserts the title equals the reference list for forward + namespaced calls (new `code_lens_resolve` harness helper);
- VS Code (`editors/vscode/src/test/codeLens.test.ts` + `testFixture/codeLensRefs.tcl`): asserts the resolved titles in-editor.

The fully namespace-correct attribution of an *ambiguous* bare call that crosses files would need workspace-wide proc knowledge in the resolver (as the Rust port has); left as a follow-up, since the reported case is single-file.


---
_Generated by [Claude Code](https://claude.ai/code)_
